src/grimp/adaptors/graph.py

```diff
@@ -129,344 +52,149 @@ def add_import(
-                children.add(potential_child)
-        return children
+        return self._rustgraph.find_children(module)
 
     def find_descendants(self, module: str) -> Set[str]:
         # It doesn't make sense to find the descendants of a squashed module, as we don't store
         # the descendants in the graph.
         if self.is_module_squashed(module):
             raise ValueError("Cannot find descendants of a squashed module.")
-
-        descendants = set()
-        for potential_descendant in self.modules:
-            if Module(potential_descendant).is_descendant_of(Module(module)):
-                descendants.add(potential_descendant)
-        return descendants
-
-    # Direct imports
-    # --------------
+        return self._rustgraph.find_descendants(module)
 
     def direct_import_exists(
         self, *, importer: str, imported: str, as_packages: bool = False
     ) -> bool:
-        if not as_packages:
-            return imported in self.find_modules_directly_imported_by(importer)
-
-        importer_modules = self._all_modules_in_package(importer)
-        imported_modules = self._all_modules_in_package(imported)
-
-        if importer_modules & imported_modules:
-            # If there are shared modules between the two, one of the modules is a descendant
-            # of the other (or they're both the same module). This doesn't make sense in
-            # this context, so raise an exception.
-            raise ValueError("Modules have shared descendants.")
-
-        # Return True as soon as we find a path between any of the modules in the subpackages.
-        for candidate_importer in importer_modules:
-            imported_by_importer = self.find_modules_directly_imported_by(candidate_importer)
-            for candidate_imported in imported_modules:
-                if candidate_imported in imported_by_importer:
-                    return True
-        return False
+        return self._rustgraph.direct_import_exists(
+            importer=importer, imported=imported, as_packages=as_packages
+        )
 
     def find_modules_directly_imported_by(self, module: str) -> Set[str]:
-        return self._importeds_by_importer[module]
+        return self._rustgraph.find_modules_directly_imported_by(module)
 
     def find_modules_that_directly_import(self, module: str) -> Set[str]:
-        return self._importers_by_imported[module]
-
-    def get_import_details(self, *, importer: str, imported: str) -> List[graph.DetailedImport]:
-        import_details_for_importer = self._import_details.get(importer, [])
-        # Only include the details for the imported module.
-        # Note: we copy each details dictionary at this point, as our deepcopying
-        # only copies the dictionaries by reference.
-        return [i.copy() for i in import_details_for_importer if i["imported"] == imported]
+        if module in self._rustgraph.get_modules():
```

Comment:
This check is costly. Let's add a benchmark and then (probably) push the check down to Rust.